File: code/backend/app/evolucion/router/evolucion.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import Date, cast
from sqlalchemy.orm import Session

from app.core.dependencies import obtener_usuario_actual
from app.evaluaciones.model.evaluacion import Evaluacion
from app.evaluaciones.schema.evaluacion import imc
from app.evolucion.schema.evolucion import (
    MetricaEvolucion,
    MostrarEvolucion,
    PacienteEvolucion,
    PeriodoEvolucion,
    PuntoEvolucion,
    ResumenEvolucion,
)
from app.pacientes.model.paciente import Paciente
from app.usuarios.model.usuario import Usuario
from database import get_db


router = APIRouter(
    prefix="/evolucion",
    tags=["Evolucion"],
    dependencies=[Depends(obtener_usuario_actual)],
)


UNIDADES = {
    MetricaEvolucion.peso: "kg",
    MetricaEvolucion.imc: "kg/m²",
    MetricaEvolucion.masa_muscular: "%",
}


def _valor_de_metrica(evaluacion: Evaluacion, metrica: MetricaEvolucion) -> float:
    if metrica == MetricaEvolucion.peso:
        return float(evaluacion.peso)
    if metrica == MetricaEvolucion.masa_muscular:
        return float(evaluacion.masa_muscular)
    return imc(evaluacion.peso, evaluacion.altura)


def _filtrar_por_periodo(
    consulta,
    periodo: PeriodoEvolucion,
    desde: date | None,
    hasta: date | None,
):
    if periodo == PeriodoEvolucion.personalizado:
        if desde is None or hasta is None:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="El periodo personalizado requiere desde y hasta",
            )
        if desde > hasta:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail="La fecha desde no puede ser posterior a hasta",
            )
        return consulta.filter(cast(Evaluacion.fecha_registro, Date) >= desde).filter(
            cast(Evaluacion.fecha_registro, Date) <= hasta
        )

    if periodo == PeriodoEvolucion.ultimo_mes:
        limite = date.today() - timedelta(days=30)
        return consulta.filter(cast(Evaluacion.fecha_registro, Date) >= limite)

    if periodo == PeriodoEvolucion.ultimos_3_meses:
        limite = date.today() - timedelta(days=90)
        return consulta.filter(cast(Evaluacion.fecha_registro, Date) >= limite)

    return consulta


@router.get("/paciente/{paciente_id}", response_model=MostrarEvolucion)
def mostrar_evolucion_paciente(
    paciente_id: int,
    metrica: MetricaEvolucion = Query(default=MetricaEvolucion.peso),
    periodo: PeriodoEvolucion = Query(default=PeriodoEvolucion.ultimas_3),
    desde: date | None = Query(default=None),
    hasta: date | None = Query(default=None),
    db: Session = Depends(get_db),
    usuario_actual: Usuario = Depends(obtener_usuario_actual),
):
    paciente = (
        db.query(Paciente)
        .filter(
            Paciente.id == paciente_id,
            Paciente.esta_activo.is_(True),
            Paciente.usuario_id == usuario_actual.id,
        )
        .first()
    )
    if paciente is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Paciente no encontrado",
        )

    consulta = db.query(Evaluacion).filter(
        Evaluacion.paciente_id == paciente_id,
        Evaluacion.esta_activo.is_(True),
    )

    if periodo == PeriodoEvolucion.ultimas_3:
        evaluaciones = (
            consulta.order_by(
                Evaluacion.fecha_registro.desc(), Evaluacion.nro_evaluacion.desc()
            )
            .limit(3)
            .all()
        )
        evaluaciones.sort(
            key=lambda evaluacion: (
                evaluacion.fecha_registro,
                evaluacion.nro_evaluacion,
            )
        )
    else:
        evaluaciones = (
            _filtrar_por_periodo(consulta, periodo, desde, hasta)
            .order_by(Evaluacion.fecha_registro.asc(), Evaluacion.nro_evaluacion.asc())
            .all()
        )

    puntos = [
        PuntoEvolucion(
            nro_evaluacion=evaluacion.nro_evaluacion,
            fecha=evaluacion.fecha_registro,
            valor=_valor_de_metrica(evaluacion, metrica),
        )
        for evaluacion in evaluaciones
    ]

    valor_inicial = puntos[0].valor if puntos else None
    valor_actual = puntos[-1].valor if puntos else None
    cambio = (
        round(valor_actual - valor_inicial, 2)
        if valor_inicial is not None and valor_actual is not None
        else None
    )

    return MostrarEvolucion(
        paciente=PacienteEvolucion(id=paciente.id, nombre=paciente.nombre),
        metrica=metrica,
        unidad=UNIDADES[metrica],
        periodo=periodo,
        resumen=ResumenEvolucion(
            valor_inicial=valor_inicial,
            valor_actual=valor_actual,
            cambio=cambio,
            cantidad_evaluaciones=len(puntos),
        ),
        puntos=puntos,
    )

File: code/frontend_flutter/lib/models/evolucion.dart
enum MetricaEvolucion { peso, imc, masaMuscular }

extension MetricaEvolucionX on MetricaEvolucion {
  String get apiValue => switch (this) {
    MetricaEvolucion.peso => 'peso',
    MetricaEvolucion.imc => 'imc',
    MetricaEvolucion.masaMuscular => 'masa_muscular',
  };

  String get etiqueta => switch (this) {
    MetricaEvolucion.peso => 'Peso',
    MetricaEvolucion.imc => 'IMC',
    MetricaEvolucion.masaMuscular => 'Masa muscular',
  };

  String get unidad => switch (this) {
    MetricaEvolucion.peso => 'kg',
    MetricaEvolucion.imc => 'kg/m\u00b2',
    MetricaEvolucion.masaMuscular => '%',
  };

  static MetricaEvolucion fromApi(String value) => switch (value) {
    'peso' => MetricaEvolucion.peso,
    'imc' => MetricaEvolucion.imc,
    'masa_muscular' => MetricaEvolucion.masaMuscular,
    _ => throw FormatException('M\u00e9trica no reconocida: $value'),
  };
}

enum PeriodoEvolucion {
  ultimas3,
  ultimoMes,
  ultimos3Meses,
  todoHistorial,
  personalizado,
}

extension PeriodoEvolucionX on PeriodoEvolucion {
  String get apiValue => switch (this) {
    PeriodoEvolucion.ultimas3 => 'ultimas_3',
    PeriodoEvolucion.ultimoMes => 'ultimo_mes',
    PeriodoEvolucion.ultimos3Meses => 'ultimos_3_meses',
    PeriodoEvolucion.todoHistorial => 'todo_historial',
    PeriodoEvolucion.personalizado => 'personalizado',
  };

  String get etiqueta => switch (this) {
    PeriodoEvolucion.ultimas3 => '\u00daltimas 3 evaluaciones',
    PeriodoEvolucion.ultimoMes => '\u00daltimo mes',
    PeriodoEvolucion.ultimos3Meses => '\u00daltimos 3 meses',
    PeriodoEvolucion.todoHistorial => 'Todo el historial',
    PeriodoEvolucion.personalizado => 'Rango personalizado',
  };

  static PeriodoEvolucion fromApi(String value) => switch (value) {
    'ultimas_3' => PeriodoEvolucion.ultimas3,
    'ultimo_mes' => PeriodoEvolucion.ultimoMes,
    'ultimos_3_meses' => PeriodoEvolucion.ultimos3Meses,
    'todo_historial' => PeriodoEvolucion.todoHistorial,
    'personalizado' => PeriodoEvolucion.personalizado,
    _ => throw FormatException('Per\u00edodo no reconocido: $value'),
  };
}

class PacienteEvolucion {
  const PacienteEvolucion({required this.id, required this.nombre});

  final int id;
  final String nombre;

  factory PacienteEvolucion.fromJson(Map<String, dynamic> json) {
    return PacienteEvolucion(
      id: json['id'] as int,
      nombre: json['nombre'] as String,
    );
  }
}

class PuntoEvolucion {
  const PuntoEvolucion({
    required this.numeroEvaluacion,
    required this.fecha,
    required this.valor,
  });

  final int numeroEvaluacion;
  final DateTime fecha;
  final double valor;

  factory PuntoEvolucion.fromJson(Map<String, dynamic> json) {
    return PuntoEvolucion(
      numeroEvaluacion: json['nro_evaluacion'] as int,
      fecha: DateTime.parse(json['fecha'] as String),
      valor: (json['valor'] as num).toDouble(),
    );
  }
}

class ResumenEvolucion {
  const ResumenEvolucion({
    required this.valorInicial,
    required this.valorActual,
    required this.cambio,
    required this.cantidadEvaluaciones,
  });

  final double? valorInicial;
  final double? valorActual;
  final double? cambio;
  final int cantidadEvaluaciones;

  factory ResumenEvolucion.fromJson(Map<String, dynamic> json) {
    double? numero(dynamic valor) => valor is num ? valor.toDouble() : null;

    return ResumenEvolucion(
      valorInicial: numero(json['valor_inicial']),
      valorActual: numero(json['valor_actual']),
      cambio: numero(json['cambio']),
      cantidadEvaluaciones: json['cantidad_evaluaciones'] as int,
    );
  }
}

class Evolucion {
  const Evolucion({
    required this.paciente,
    required this.metrica,
    required this.unidad,
    required this.periodo,
    required this.resumen,
    required this.puntos,
  });

  final PacienteEvolucion paciente;
  final MetricaEvolucion metrica;
  final String unidad;
  final PeriodoEvolucion periodo;
  final ResumenEvolucion resumen;
  final List<PuntoEvolucion> puntos;

  factory Evolucion.fromJson(Map<String, dynamic> json) {
    return Evolucion(
      paciente: PacienteEvolucion.fromJson(
        json['paciente'] as Map<String, dynamic>,
      ),
      metrica: MetricaEvolucionX.fromApi(json['metrica'] as String),
      unidad: json['unidad'] as String,
      periodo: PeriodoEvolucionX.fromApi(json['periodo'] as String),
      resumen: ResumenEvolucion.fromJson(
        json['resumen'] as Map<String, dynamic>,
      ),
      puntos: (json['puntos'] as List<dynamic>)
          .map(
            (punto) => PuntoEvolucion.fromJson(punto as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}
